src/TabComplete.js

```diff
@@ -45,21 +48,32 @@ class TabComplete {
         this.isFirstWord = false; // true if you tab-complete on the first word
         this.enterTabCompleteTimerId = null;
         this.inPassiveMode = false;
+        this.memberTabOrder = {};
+        this.memberOrderSeq = 0;
     }
 
     /**
-     * @param {Entry[]} completeList
+     * Call this when a a UI element representing a tab complete entry has been clicked
+     * @param {entry} The entry that was clicked
      */
-    setCompletionList(completeList) {
-        this.list = completeList;
+    onEntryClick(entry) {
         if (this.opts.onClickCompletes) {
-            // assign onClick listeners for each entry to complete the text
-            this.list.forEach((l) => {
-                l.onClick = () => {
-                    this.completeTo(l);
-                }
-            });
+            this.completeTo(entry);
+        }
+    }
+
+    loadEntries(room) {
+        this._makeEntries(room);
+        this._initSorting(room);
+        this._sortEntries();
+    }
+
+    onMemberSpoke(member) {
+        if (this.memberTabOrder[member.userId] === undefined) {
+            this.list.push(new MemberEntry(member));
         }
+        this.memberTabOrder[member.userId] = this.memberOrderSeq++;
+        this._sortEntries();
```

Comment:
Do we *really* have to go through and re-sort the entire member list when someone speaks? I thought the whole point of this was to avoid doing that, and yet here we are. `onMemberSpoke` is called when `RoomView.onRoomTimeline` is called, and we don't save ourselves a sort operation (we know that member should be at the head of the list; they just spoke!). Likewise, `_sortEntries` doesn't appear to do any optimisations beyond saving ourselves the O(N) scan of the timeline which was done in `_initSorting`.